linkerd/stack/src/watch.rs

```diff
@@ -3,37 +3,36 @@ use std::task::{Context, Poll};
 use tokio::sync::watch;
 use tracing::Instrument;
 
-/// Defines how to update a [`Service`] in the background as a [`watch::Receiver`].
-/// containing a `T-`typed target is updated.
-pub trait UpdateWatch<T>: Send + 'static {
-    /// The type of service built based on the watched target.
-    type Service: Clone + Default + Send + Sync + 'static;
-
-    /// Update the inner service with a new `target` value.
-    ///
-    /// If the inner service has not changed as a result of the update, this
-    /// method may return `None`. If `None` is returned, no update will be
-    /// published to the [`SpawnWatch`] service.
-    ///
-    /// This method is also called initially when the first [`SpawnWatch`]
-    /// service is constructed. In this case, it should generally not return
-    /// `None`, but if it does, the [`Default`] value of the [`Self::Service`]
-    /// type is used.
-    fn update(&mut self, target: &T) -> Option<Self::Service>;
-}
-
-/// Builds [`SpawnWatch`] services, where an inner [`Service`] is updated
-/// dynamically by a background task as a target wrapped in a
-/// [`watch::Receiver`] changes.
+/// Builds [`SpawnWatch`] services from a target that produces a `P`-typed
+/// [`watch::Receiver`].
 ///
-/// `N` is a [`NewService`] which produces types implementing [`UpdateWatch`],
-/// which define the logic for how to update the inner [`Service`].
+/// A background task is spawned that updates the inner service when the watch
+/// value changes.
 #[derive(Clone, Debug)]
 pub struct NewSpawnWatch<P, N> {
-    new_update: N,
+    inner: N,
     _marker: std::marker::PhantomData<fn(P)>,
 }
 
+/// Builds `NewFromTuple` instances within a `NewSpawnWatchStack`.
+///
+/// When we use a `T`-typed target to produce a `U`-typed `watch::Receiver`,
+/// this module converts the `U`-typed target value to a `P`-typed target for
+/// the inner stack. `P` must implement `From<(U, T)>`.
+#[derive(Clone, Debug)]
+pub struct NewWatchedFromTuple<P, N> {
+    inner: N,
+    _marker: std::marker::PhantomData<fn() -> P>,
+}
+
+/// Builds services by passing `P` typed values to an `N`-typed inner stack.
+#[derive(Clone, Debug)]
```

Comment:
these `Clone` and `Debug` impls will require that `P` be `Clone`/`Debug` --- do we need manual impls, or is that not a problem here?

Reply:
I think practically we always expect targets are clone but you are right that it would be more correcter to manual impl